### PR TITLE
[MO] - [oauth + clients] -> removing oauth external lister dependency + refactor clients

### PR DESCRIPTION
**Pull request by see-quick**
Signed-off-by: morsak <xorsak02@stud.fit.vutbr.cz>

### Type of change

- Enhancement / new feature
- Refactoring

### Description

This PR removing the dependency on external listeners in ouath test suite. Moreover, refactored the current client's resource which currently has this structure.

- KafkaClientsResource (abstract)
  - KafkaBasicClientResource
     - KafkaOauthClientsResource 
  - KafkaBridgeClientsResource
  - KafkaTracingClientsResource

I (@Frawless ) did a different part of this PR than @see-quick so I approved his changes (I already tested them).

### Checklist

- [ ] Make sure all tests pass

### CODE REVIEW

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=oauth test_only

**Comment by see-quick** [comment]
@strimzi-ci run tests profile=oauth test_only

**Comment by Frawless** [comment]
@strimzi-ci run tests profile=oauth test_only

**Comment by Frawless** [comment]
@strimzi-ci run tests profile=oauth test_only

**Review by Frawless — APPROVED** [review]

**Comment by Frawless** [comment]
@strimzi-ci run tests profile=regression testcase=RecoveryST#testRecoveryFromKafkaStatefulSetDeletion  test_only

File: systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthAbstractST.java [hunk 1]
```diff
@@ -72,73 +67,31 @@ void setup() throws Exception {
...
         createSecretsForDeployments();
+    }
+
+    @AfterEach
+    void tearDown() {
 
-        KafkaUserResource.tlsUser(CLUSTER_NAME, OAUTH_CLIENT_NAME).done();
+        for (Job job : kubeClient().getClient().batch().jobs().inNamespace(NAMESPACE).list().getItems()) {
+            LOGGER.info("Deleting {} job", job.getMetadata().getName());
+            kubeClient().getClient().batch().jobs().inNamespace(NAMESPACE).delete(job);
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthAbstractST.java:82]: Why not to use the `JobUtils.deleteJob()`?

File: systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthAuthorizationST.java [hunk 2]
```diff
@@ -222,76 +231,70 @@ void testClusterVerification() {
...
+                TEAM_B_PRODUCER_NAME,
+                TEAM_B_CONSUMER_NAME,
+                KafkaResources.tlsBootstrapAddress(CLUSTER_NAME),
+                TOPIC_A,
+                MESSAGE_COUNT,
+                "",
+                "x-" + ClientUtils.generateRandomConsumerGroup(),
+                TEAM_B_CLIENT,
+                TEAM_B_CLIENT_SECRET,
+                keycloakInstance.getOauthTokenEndpointUri());
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthAuthorizationST.java:298]: Another suggestion -> shouldn't be better to have builder for this? Just asking ... for the future PR?

File: systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthPlainST.java [hunk 3]
```diff
@@ -187,19 +184,35 @@ void testProducerConsumerMirrorMaker() {
         TestUtils.waitFor("Waiting for Mirror Maker will copy messages from " + CLUSTER_NAME + " to " + targetKafkaCluster,
             Constants.GLOBAL_CLIENTS_POLL, Constants.TIMEOUT_FOR_MIRROR_MAKER_COPY_MESSAGES_BETWEEN_BROKERS,
             () -> {
-                oauthExternalKafkaClient.setConsumerGroup(ClientUtils.generateRandomConsumerGroup());
-                return oauthExternalKafkaClient.receiveMessagesPlain() == MESSAGE_COUNT;
+                LOGGER.info("Deleting the Job");
+                kubeClient().getClient().batch().jobs().inNamespace(NAMESPACE).withName(OAUTH_CONSUMER_NAME).delete();
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthPlainST.java:188]: Same as above: `Why not to use the JobUtils.deleteJob()?`

File: systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthPlainST.java [hunk 4]
```diff
@@ -273,19 +286,35 @@ void testProducerConsumerMirrorMaker2() {
         TestUtils.waitFor("Waiting for Mirror Maker 2 will copy messages from " + kafkaSourceClusterName + " to " + kafkaTargetClusterName,
             Duration.ofSeconds(30).toMillis(), Constants.TIMEOUT_FOR_MIRROR_MAKER_COPY_MESSAGES_BETWEEN_BROKERS,
             () -> {
-                oauthExternalKafkaClient.setConsumerGroup(ClientUtils.generateRandomConsumerGroup());
-                return oauthExternalKafkaClient.receiveMessagesPlain() == MESSAGE_COUNT;
+                LOGGER.info("Deleting the Job {}", OAUTH_CONSUMER_NAME);
+                kubeClient().getClient().batch().jobs().inNamespace(NAMESPACE).withName(OAUTH_CONSUMER_NAME).delete();
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthPlainST.java:290]: Same as above

File: systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthAbstractST.java [hunk 5]
```diff
@@ -72,73 +67,31 @@ void setup() throws Exception {
...
+        }
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthAbstractST.java:83]: Why not to use the `JobUtils.deleteJob()`?

And question -> don't we want to add something like `JobsClient()` to not always do the `kubeClient().getClient().batch().jobs()`? This is just a question/suggestion.

File: systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthAuthorizationST.java (hunk 2 above)
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthAuthorizationST.java:298]: Another suggestion -> shouldn't be better to have builder for this? Just asking ... for the future PR?

File: systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthPlainST.java (hunk 3 above)
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthPlainST.java:188]: Same as above: Why not to use the JobUtils.deleteJob()?

File: systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthPlainST.java (hunk 4 above)
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthPlainST.java:290]: Same as above

**Review by im-konge — COMMENTED** [review]
Some minor suggestions/nits/questions, but you're done a really good work.

File: systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthAbstractST.java (hunk 5 above)
Comment (Frawless) [reply systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthAbstractST.java:83]: Leftover, feel free to open issue to add this job implementation to kubeclient.

File: systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthAuthorizationST.java (hunk 2 above)
Comment (Frawless) [reply systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthAuthorizationST.java:298]: Open issue for that. We discuss it and we decide it needs separete PR.

Comment (im-konge) [reply systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthAuthorizationST.java:298]: Thanks :) gonna open it.

File: systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthAbstractST.java (hunk 5 above)
Comment (im-konge) [reply systemtest/src/test/java/io/strimzi/systemtest/security/oauth/OauthAbstractST.java]: Perfect, thanks

**Review by im-konge — APPROVED** [review]
LGTM :)

File: systemtest/src/main/java/io/strimzi/systemtest/utils/ClientUtils.java [hunk 6]
```diff
@@ -60,6 +60,12 @@ public static void waitForClientSuccess(String jobName, String namespace, int me
             () -> kubeClient().getClient().batch().jobs().inNamespace(namespace).withName(jobName).get().getStatus().getSucceeded().equals(1));
     }
 
+    public static void waitForClientFailure(String jobName, String namespace, long timeout) {
+        LOGGER.info("Waiting for producer/consumer:{} will be in error state", jobName);
+        TestUtils.waitFor("job finished", Constants.GLOBAL_POLL_INTERVAL, timeout,
+            () -> kubeClient().getClient().batch().jobs().inNamespace(namespace).withName(jobName).get().getStatus().getSucceeded().equals(1));
+    }
```
Comment (scholzj) [inline systemtest/src/main/java/io/strimzi/systemtest/utils/ClientUtils.java:67]: Should this be a `waitForJobFailure` and sit in `JobUtils`? apart from the log message, there is nothing client specific.

File: systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeKafkaExternalListenersST.java [hunk 7]
```diff
@@ -164,5 +166,8 @@ private void testWeirdUsername(String weirdUserName, KafkaListenerAuthentication
     @BeforeAll
     void createClassResources() throws Exception {
         deployClusterOperator(NAMESPACE);
+
+        kafkaBridgeClientJob = new KafkaBridgeClientsResource(producerName, consumerName, KafkaBridgeResources.serviceName(CLUSTER_NAME),
+            TOPIC_NAME, MESSAGE_COUNT, "", ClientUtils.generateRandomConsumerGroup(), bridgePort, 1000, 1000);
```
Comment (scholzj) [inline systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeKafkaExternalListenersST.java:171]: This seems to be the same for all the Bridge tests. Can this be moved in some way to the abstract class?

File: systemtest/src/test/java/io/strimzi/systemtest/specific/ClusterOperationST.java [hunk 8]
```diff
@@ -32,16 +32,19 @@
 public class ClusterOperationST extends AbstractST {
 
     private static final Logger LOGGER = LogManager.getLogger(ClusterOperationST.class);
+    private KafkaBasicClientResource kafkaBasicClientResource;
```
Comment (scholzj) [inline systemtest/src/test/java/io/strimzi/systemtest/specific/ClusterOperationST.java:35]: What is the value of having this at the class level? Looks like you use it only within a single loop.

**Review by scholzj — APPROVED** [review]
Looks good overall ... just some minor comments.

**Comment by Frawless** [comment]
@strimzi-ci run tests profile=bridge test_only

**Comment by Frawless** [comment]
@strimzi-ci run tests profile=bridge testcase=HttpBrdige.*ST test_only

**Comment by Frawless** [comment]
@strimzi-ci run tests profile=bridge testcase=HttpBridge.*ST test_only

**Comment by Frawless** [comment]
@strimzi-ci run tests profile=bridge test_only

**Pull request merged**